### PR TITLE
Linter 130s timeout and primitive logging

### PR DESCRIPTION
**Pull request by andrlime**
Added 130s timeout. After 130s, if the process has not returned 0, force push a failure to Firebase. Moved some log_e and log_i to the ss stream which theoretically goes to Firebase too.

needs more testing lol

### CODE REVIEW

**Review by egelja — CHANGES_REQUESTED** [review]
If you are hoisting up the fork you should also hoist up the corresponding if statement, rather than passing around a PID

File: linter/src/crow/crow.cpp [hunk 1]
```diff
@@ -1,5 +1,9 @@
 #include "crow.hpp"
 
+#include "firebase/fetching.hpp"
+
+#include <sys/wait.h>
```
Comment (egelja) [inline linter/src/crow/crow.cpp:5]: Delete pls

File: linter/src/firebase/fetching.hpp [hunk 2]
```diff
@@ -34,6 +34,9 @@ void set_lint_success(
 );
 
 std::optional<std::string> get_algo(const std::string& uid, const std::string& algo_id);
+std::optional<std::string> get_algo_status(
+    const std::string& uid, const std::string& algo_id
+); // failure, pending, success
```
Comment (egelja) [inline linter/src/firebase/fetching.hpp:39]: Shouldn’t we return an enumeration instead?

File: linter/src/spawner/main.cpp [hunk 3]
```diff
@@ -76,29 +76,73 @@ main(int argc, const char** argv)
...
+        std::optional<std::string> linting_status =
+            nutc::client::get_algo_status(uid, algoid);
+
+        // If status is pending, force push a failure
+        if (linting_status.has_value()) {
+            if (linting_status == "pending") {
+                // Push failure
+                std::string error_msg =
+                    "unknown runtime error: your code is syntactically correct but "
+                    "crashed during runtime";
```
Comment (egelja) [inline linter/src/spawner/main.cpp:110]: Times out

File: linter/src/spawner/main.cpp [hunk 4]
```diff
@@ -76,29 +76,73 @@ main(int argc, const char** argv)
...
+    // Secondary watchdog to check for crash/pending after 130s
+    std::thread check_pending_thread([&ss = ss, &algoid = algoid, &uid = uid]() {
+        std::this_thread::sleep_for(std::chrono::seconds(130));
+
+        // Check status from Firebase
+        std::optional<std::string> linting_status =
+            nutc::client::get_algo_status(uid, algoid);
+
+        // If status is pending, force push a failure
+        if (linting_status.has_value()) {
```
Comment (egelja) [inline linter/src/spawner/main.cpp:105]: Combine into one if

**Review by egelja — CHANGES_REQUESTED** [review]
Looks much better

**Review by stevenewald — CHANGES_REQUESTED** [review]

File: linter/src/crow/crow.cpp [hunk 5]
```diff
@@ -24,9 +26,50 @@ get_server_thread()
             std::string uid = req.url_params.get("uid");
             std::string algo_id = req.url_params.get("algo_id");
 
+            // Watchdog to check for crash/pending after 130s
+            std::thread check_pending_thread([&algo_id = algo_id, &uid = uid]() {
+                std::this_thread::sleep_for(std::chrono::seconds(130));
```
Comment (egelja) [inline linter/src/crow/crow.cpp:31]: Can you make the timeout configurable in `config.h.in`?

File: linter/src/crow/crow.cpp [hunk 6]
```diff
@@ -24,9 +26,50 @@ get_server_thread()
...
+                // Check status from Firebase
+                nutc::client::LintingResultOption linting_status =
+                    nutc::client::get_algo_status(uid, algo_id);
+
+                // If status is pending, force push a failure
+                if (linting_status == nutc::client::LintingResultOption::PENDING) {
+                    // Push failure
+                    std::string error_msg =
+                        "unknown runtime error: your code is syntactically correct but "
+                        "timed out (130 seconds) while linting";
```
Comment (egelja) [inline linter/src/crow/crow.cpp:42]:
```suggestion
                        "timed out after 130 seconds while linting";
```

also will need to make this time configurable

File: linter/src/crow/crow.cpp [hunk 7]
```diff
@@ -24,9 +26,50 @@ get_server_thread()
...
+
+                    nutc::client::set_lint_result(uid, algo_id, false);
+                    nutc::client::set_lint_failure(uid, algo_id, error_msg);
+
+                    log_e(
+                        main,
+                        "Algoid {} for uid {} still pending after 130s. FORCE PUSHING "
+                        "failure to "
+                        "Firebase.",
```
Comment (egelja) [inline linter/src/crow/crow.cpp:51]:
```suggestion
                        "failure to Firebase.",
```

File: linter/src/crow/crow.cpp [hunk 8]
```diff
@@ -24,9 +26,50 @@ get_server_thread()
...
+                        algo_id,
+                        uid
+                    );
+                }
+
+                if (linting_status == nutc::client::LintingResultOption::UNKNOWN) {
+                    // can add a push to firebase here
+                    log_e(
+                        main,
+                        "Algoid {} for uid {} unknown status after 130s.",
```
Comment (egelja) [inline linter/src/crow/crow.cpp:61]:
```suggestion
                        "Algo id {} for uid {} has unknown status after 130s.",
```

File: linter/src/crow/crow.cpp [hunk 9]
```diff
@@ -24,9 +26,50 @@ get_server_thread()
...
+                        main,
+                        "Algoid {} for uid {} still pending after 130s. FORCE PUSHING "
+                        "failure to "
+                        "Firebase.",
+                        algo_id,
+                        uid
+                    );
+                }
+
+                if (linting_status == nutc::client::LintingResultOption::UNKNOWN) {
```
Comment (egelja) [inline linter/src/crow/crow.cpp:57]: you should use a switch here

File: linter/src/firebase/fetching.cpp [hunk 10]
```diff
@@ -148,6 +148,49 @@ get_algo(const std::string& uid, const std::string& algo_id)
...
+
+    if (linting_result == "failure") {
+        return nutc::client::LintingResultOption::FAILURE;
+    }
+    else if (linting_result == "success") {
+        return nutc::client::LintingResultOption::SUCCESS;
+    }
+    else {
+        return nutc::client::LintingResultOption::PENDING;
+    }
```
Comment (egelja) [inline linter/src/firebase/fetching.cpp:191]: do a switch on `linting_result[0]` with some asserts, its faster

File: linter/src/crow/crow.cpp [hunk 11]
```diff
@@ -24,9 +26,50 @@ get_server_thread()
...
+                    // can add a push to firebase here
+                    log_e(
+                        main,
+                        "Algoid {} for uid {} unknown status after 130s.",
+                        algo_id,
+                        uid
+                    );
+                }
+
+                std::exit(1);
```
Comment (egelja) [inline linter/src/crow/crow.cpp:67]: doesn't this exit the full linter?

File: linter/src/spawner/main.cpp [hunk 12]
```diff
@@ -76,29 +76,38 @@ main(int argc, const char** argv)
     std::stringstream ss;
 
     // Watchdog to kill after 120s
-    std::thread timeout_thread([&ss, &algoid = algoid, &uid = uid]() {
+    std::thread timeout_thread([&ss = ss, &algoid = algoid, &uid = uid]() {
         std::this_thread::sleep_for(std::chrono::seconds(120));
```
Comment (egelja) [inline linter/src/spawner/main.cpp:80]: this should be configurable too

File: linter/src/spawner/main.cpp [hunk 13]
```diff
@@ -76,29 +76,38 @@ main(int argc, const char** argv)
...
+
         log_e(
             main,
             "Timeout reached. Exiting process. Failed lint for algoid {} and uid {}.",
             algoid,
             uid
         );
-        nutc::client::set_lint_failure(uid, algoid, ss.str() + "Failure\n");
+        ss << "[linter] FAILED to lint algo_id " << algoid << " for uid " << uid
+           << "\n";
```
Comment (egelja) [inline linter/src/spawner/main.cpp:89]: You should use fmt lib --- its already a dep

File: linter/src/spawner/main.cpp [hunk 14]
```diff
@@ -76,29 +76,38 @@ main(int argc, const char** argv)
...
+
+        nutc::client::set_lint_failure(uid, algoid, ss.str() + "Failure!\n");
         std::exit(1);
     });
     timeout_thread.detach();
 
     // Log this event
     log_i(main, "Linting algo_id: {} for user: {}", algoid, uid);
+    ss << "[linter] starting to lint algo_id " << algoid << " for uid " << uid << "\n";
```
Comment (egelja) [inline linter/src/spawner/main.cpp:98]: You should use fmt lib --- its already a dep

File: linter/src/spawner/main.cpp [hunk 15]
```diff
@@ -76,29 +76,38 @@ main(int argc, const char** argv)
...
     // Initialize py
     pybind11::initialize_interpreter();
 
     // Actually lint the algo and set success
-    std::string response = nutc::lint::lint(uid, algoid);
-    nutc::client::set_lint_success(uid, algoid, ss.str() + "Success\n");
+    std::string response = nutc::lint::lint(uid, algoid, ss);
+
     log_i(main, "Finished linting algo_id: {} for user: {}", algoid, uid);
+    ss << "[linter] exited linting process and finished linting algo_id " << algoid
```
Comment (egelja) [inline linter/src/spawner/main.cpp:107]: You should use fmt lib --- its already a dep

**Review by egelja — CHANGES_REQUESTED** [review]
looks good

File: linter/src/firebase/fetching.cpp (hunk 10 above)
Comment (andrlime) [reply linter/src/firebase/fetching.cpp]: i didn't do asserts here, was not sure what to assert

**Review by stevenewald — APPROVED** [review]

**Pull request merged**